Support ETM (Encrypt-then-MAC) variants for HMAC

scott-xu opened the pull request:
Should fix #875 and fix #1050 and fix #837 and fix #1092

References:
1. https://github.com/openssh/openssh-portable/blob/cbbdf868bce431a59e2fa36ca244d5739429408d/PROTOCOL
2. https://github.com/openssh/openssh-portable/blob/cbbdf868bce431a59e2fa36ca244d5739429408d/packet.c

Written by hand **without testing** by now.

scott-xu commented:
Would you mind giving a try for this PR?
Note: it is written without testing at this moment. Feel free to improve the PR if you want.
@PashCracken @SkymaX86 @robert-scheck @alef75 @Jedrzej94 @Ph43n0m @AliceTJ

Rob-Hague commented:
Nice. Testing should be straightforward:

```diff
diff --git a/test/Renci.SshNet.IntegrationTests/HmacTests.cs b/test/Renci.SshNet.IntegrationTests/HmacTests.cs
index 993e5ec98..e5057248c 100644
--- a/test/Renci.SshNet.IntegrationTests/HmacTests.cs
+++ b/test/Renci.SshNet.IntegrationTests/HmacTests.cs
@@ -58,6 +58,24 @@ public void HmacSha2_512()
             DoTest(MessageAuthenticationCodeAlgorithm.HmacSha2_512);
         }
 
+        [TestMethod]
+        public void HmacSha1Etm()
+        {
+            DoTest(MessageAuthenticationCodeAlgorithm.HmacSha1Etm);
+        }
+
+        [TestMethod]
+        public void HmacSha2_256_Etm()
+        {
+            DoTest(MessageAuthenticationCodeAlgorithm.HmacSha2_256_Etm);
+        }
+
+        [TestMethod]
+        public void HmacSha2_512_Etm()
+        {
+            DoTest(MessageAuthenticationCodeAlgorithm.HmacSha2_512_Etm);
+        }
+
         private void DoTest(MessageAuthenticationCodeAlgorithm macAlgorithm)
         {
             _remoteSshdConfig.ClearMessageAuthenticationCodeAlgorithms()
```

Rob-Hague commented on src/Renci.SshNet/Session.cs:
```diff
@@ -1257,6 +1279,20 @@ private Message ReceiveMessage(Socket socket)
+            // validate encrypted message against MAC
+            if (_serverMac != null && _serverMac.ETM)
+            {
+                var clientHash = _serverMac.HashAlgorithm.ComputeHash(data, blockSize, data.Length - blockSize - serverMacLength);
```
```suggestion
                var clientHash = _serverMac.HashAlgorithm.ComputeHash(data, 0, data.Length - serverMacLength);
```

scott-xu committed:
Support ETM (Encrypt-then-MAC) variants for HMAC

scott-xu committed:
Merge branch 'sshnet:develop' into etm

scott-xu committed:
Support ETM (Encrypt-then-MAC) variants for HMAC

scott-xu committed:
Merge branch 'etm' of https://github.com/scott-xu/SSH.NET into etm

scott-xu commented:
Do we want to support `HmacSha1Etm`? IMHO, `HmacSha2_256_Etm` and `HmacSha2_512_Etm` is sufficient (and secure). @Rob-Hague

Rob-Hague commented:
I think we might as well, while we still have `hmac-sha1`

Rob-Hague commented on src/Renci.SshNet/Security/Cryptography/HMAC.cs:
```diff
@@ -0,0 +1,49 @@
+    /// <summary>
+    /// Represents the info for Message Authentication Code (MAC).
+    /// </summary>
+    public sealed class HMAC : IDisposable
```
The new type seems unnecessary. Could we just change `IKeyExchange.Create[..]Hash` to `Create[..]Hash(out bool isEncryptThenMac)` or to return a tuple e.g. `(HashAlgorithm, bool) Create[..]Hash()`?

scott-xu replied:
My first revision (not committed) is `Create[...]Hash(out bool etm)`. However, compared with other argumentless `Create[...]()` methods in `IKeyExchange` interface, it looks a little bit different.

Then I took a look at other implementations, like [OpenSSH](https://github.com/openssh/openssh-portable) and [sshj](https://github.com/hierynomus/sshj). They both have dedicated data structure for MAC. Their implementation defines the detailed hash related methods plus `etm` or `isEtm` property.

https://github.com/openssh/openssh-portable/blob/master/mac.c
https://github.com/hierynomus/sshj/blob/master/src/main/java/net/schmizz/sshj/transport/mac/MAC.java

Rob-Hague replied:
Right, but we already have HashInfo, which is practically the same thing. We don't need another type representing the same information.

scott-xu replied:
From my understanding, the `HashInfo` is kind of the factory. The proposed `HMAC` is the _product_. I couldn't find a way to merge them. Should we merge the two?

WojciechNagorski replied:
How about removing the HMAC class and move ETM information to HashInfo?
You've already added the property here:
https://github.com/sshnet/SSH.NET/pull/1316/files#diff-8ad50217d0fd4634eedc20ae503cfa6584eb26b59c5922c807b581c2455f4e25R25-R28

This would protect us from the breaking change from my second comment.

scott-xu replied:
That's a good spot! The `ETM` property was added previously and then forgot to delete.
At the very beginning, I had the very similar thoughts with you which is to leverage `HashInfo` to store `ETM` Boolean value. However, later on I found the `HashInfo` is kind of factory. The `HashAlgorithm` is the _product_.

See https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Security/IKeyExchange.cs#L72 and https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Security/IKeyExchange.cs#L80

The `HashAlgorithm` is a .NET build-in class which cannot add new properties. The only way is to change `IKeyExchange.Create[..]Hash` to `Create[..]Hash(out bool isEncryptThenMac)` (which is my first revision) or to return a tuple e.g. `(HashAlgorithm, bool) Create[..]Hash()` just as @Rob-Hague pointed. I've already explained the reason why I abandoned my first revision and created dedicated `HMAC` class. See my previous comments.

I know it is a breaking change, but I don't think it will bring critical impact. There was similar successful breaking change previously. See https://github.com/sshnet/SSH.NET/commit/24838e6173e2a7d4cbb498ce5edc3e32b254ca11

Rob-Hague replied:
I do think we should either merge the classes or go with the `out bool` parameter in `IKeyExchange`, and I would suggest doing the latter. It is a much simpler change and creates one small break in the internals of the library where it matters a lot less.

scott-xu replied:
The `Cipher` returned from `CreateServerCipher` and `CreateClientCipher` is defined in SshNet. The `Compressor` returned from `CreateCompressor` and `CreateDecompressor` is also defined in SshNet.
The `HashAlgorithm` returned from `CreateServerHash` and `CreateClientHash` is a BCL class which prevents us from extending it.
There's already a bunch of `HMAC[...]` classes in SshNet, which has the same name with BCL `HMAC[...]`, specifically used for SSH purpose.

Rob-Hague commented on src/Renci.SshNet/Messages/Message.cs:
```diff
@@ -37,7 +37,7 @@ protected override void WriteBytes(SshDataStream stream)
             base.WriteBytes(stream);
         }
 
-        internal byte[] GetPacket(byte paddingMultiplier, Compressor compressor)
```
fyi @jacobslusser (or anyone else interested), the Message->Socket logic (in `Session.SendMessage` and `Message.GetPacket`) is somewhat cumbersome. There is probably some efficiency to be gained in untangling it.

Rob-Hague commented:
Looks good to me otherwise

WojciechNagorski commented on src/Renci.SshNet/HashInfo.cs:
```diff
@@ -20,17 +20,22 @@ public class HashInfo
         /// <summary>
         /// Gets the cipher.
         /// </summary>
-        public Func<byte[], HashAlgorithm> HashAlgorithm { get; private set; }
```
Breaking change

scott-xu committed:
Remove `ETM` property from `HashInfo`

scott-xu committed:
Add support for HmacMd5Etm, HmacMd5_96_Etm, HmacSha1Etm and HmacSha1_96_Etm
Explicitly specify etm even if false

scott-xu committed:
Merge branch 'develop' into etm

scott-xu committed:
Add Encrypt-then-MAC variants to README.md

scott-xu committed:
Merge branch 'etm' of https://github.com/scott-xu/SSH.NET into etm

scott-xu committed:
Add back empty span to prevent auto link

WojciechNagorski commented:
If it weren't a breaking change, I would approve it. I need some time to think it over and check it out. 

@Rob-Hague any idea?

Rob-Hague commented:
Yes I think we should go with the bool property on HashInfo and change `IKeyExchange.CreateServerHash()` from `HashAlgorithm CreateServerHash()` to `HashAlgorithm CreateServerHash(out bool isEncryptThenMac)`. I don't see the problem with that.

scott-xu commented:
Sorry I don't understand why you against `HMAC` so much.

The `HashAlgorithm CreateServerHash(out bool isEncryptThenMac)` is a break change anyway.

Shouldn't we focus on the implementation details of ETM? That's the main purpose of this PR.

scott-xu commented:
> I know everyone else's code always smells, but enough is enough - this is a collaboration, not a competition.

-- @zybexXL

Rob-Hague commented:
I think we've made it pretty clear that it's an unnecessary breaking change

`IKeyExchange` takes a break either way. That's fine: it's in the internals of the library and less likely to affect anyone.

Breaking HashInfo does not need to happen. It is much closer to the common API that people use.

You said you had already gone in this direction in an earlier iteration. I've not seen any reason to stray from that.

> Shouldn't we focus on the implementation details of ETM? That's the main purpose of this PR.

I've checked the details. They look good.

scott-xu commented:
I've already explained why I abandoned the first iteration. I don't think `HashInfo` is widely used externally just like you believe `IKeyExchange` is not widely used externally.

scott-xu commented:
To get the ball roll, someone has to compromise. Let me update the PR.

scott-xu committed:
Store ETM in `HashInfo`;
Change `HMAC Create[...]Hash()` to `HashAlgorithm Create[...]Hash(out bool isEncryptThenMAC)`

Rob-Hague reviewed (approved):
Thanks for making the change. It looks great to me. Another big pain point of the library bites the dust 🙂

Perhaps you could update the description to close a bunch of issues

scott-xu commented:
> Thanks for making the change. It looks great to me. Another big pain point of the library bites the dust 🙂
> 
> Perhaps you could update the description to close a bunch of issues

The description has had the format to close the 2 related issues. Are there more?

Rob-Hague commented:
Only one more as it turned out, I thought there were a lot more. I've updated the description

WojciechNagorski commented:
@scott-xu thanks!

scott-xu commented:
Cheers!
